sky/skylet/providers/kubernetes/node_provider.py

```diff
@@ -225,6 +233,129 @@ def _raise_pod_scheduling_errors(self, new_nodes):
+                        all_pods_scheduled = False
+                        break
+
+                    for container_status in pod.status.container_statuses:
+                        # If the container wasn't in 'ContainerCreating'
+                        # state, then we know pod wasn't scheduled or
+                        # had some other error, such as image pull error.
+                        # See list of possible reasons for waiting here:
+                        # https://stackoverflow.com/a/57886025
+                        waiting = container_status.state.waiting
+                        if waiting and waiting.reason != 'ContainerCreating':
+                            all_pods_scheduled = False
+                            break
+
+            if all_pods_scheduled:
+                return
+            time.sleep(1)
+
+        # Handle pod scheduling errors
+        try:
+            self._raise_pod_scheduling_errors(new_nodes)
+        except config.KubernetesError:
+            raise
+        except Exception as e:
+            raise config.KubernetesError(
+                'An error occurred while trying to fetch the reason '
+                'for pod scheduling failure. '
+                f'Error: {common_utils.format_exception(e)}') from None
+
+    def _wait_for_pods_to_run(self, new_nodes):
+        """Wait for pods and their containers to be ready.
+        
+        Pods may be pulling images or may be in the process of container
+        creation.
+        """
+        while True:
+            all_pods_running = True
+            # Iterate over each pod to check their status
+            for node in new_nodes:
+                pod = kubernetes.core_api().read_namespaced_pod(
+                    node.metadata.name, self.namespace)
+
+                # Continue if pod and all the containers within the
+                # pod are succesfully created and running.
+                if pod.status.phase == 'Running' \
+                    and all([container.state.running
+                             for container in pod.status.container_statuses]):
+                    continue
+
+                all_pods_running = False
+                if pod.status.phase == 'Pending':
+                    # Iterate over each container in pod to check their status
+                    for container_status in pod.status.container_statuses:
+                        waiting = container_status.state.waiting
+                        if waiting and (waiting.reason == 'ErrImagePull' or
+                                        waiting.reason == 'ImagePullBackOff'):
+                            raise config.KubernetesError(
+                                'Failed to pull docker image while '
+                                'launching the node. Please check '
+                                'your network connection. Error details: '
+                                f'{container_status.state.waiting.message}.')
```

Comment:
Once this is moved to L265, is it possible to also handle any other waiting errors (e.g., `InvalidImageName`, `CreateContainerError`, `CrashLoopBackoff`) and surface them to the user?

Reply:
Updated with a more thorough list